python/lsst/daf/butler/_utilities/thread_safe_cache.py

```diff
@@ -0,0 +1,78 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This software is dual licensed under the GNU General Public License and also
+# under a 3-clause BSD license. Recipients may choose which of these licenses
+# to use; please see the files gpl-3.0.txt and/or bsd_license.txt,
+# respectively.  If you choose the GPL option then the following text applies
+# (but note that there is still no warranty even if you opt for BSD instead):
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import threading
+from typing import Generic, TypeVar
+
+TKey = TypeVar("TKey")
+TValue = TypeVar("TValue")
+
+
+class ThreadSafeCache(Generic[TKey, TValue]):
+    """A simple thread-safe cache.  Ensures that once a value is stored for
+    a key, it does not change.
+    """
+
+    def __init__(self) -> None:
+        # This mutex may not actually be necessary for the current version of
+        # CPython, because both the 'get' and 'set' operation use underlying C
+        # functions that don't drop the GIL (at first glance anyway.) This is
+        # not documented anywhere that I can find, however, so better safe than
+        # sorry.
```

Comment:
Extremely late comment - all built-in Python types are guaranteed to be thread-safe in CPython, there is a comment about it in https://docs.python.org/3/glossary.html#term-global-interpreter-lock. But an explicit mutex may be useful once we switch to GIL-free mode 😀